compiler/qsc_ast/src/ast.rs

```diff
@@ -1416,18 +1416,33 @@ impl Display for QubitInitKind {
 }
 
 /// A path to a declaration.
-#[derive(Clone, Debug, Eq, Hash, PartialEq, Default)]
+#[derive(Clone, Debug, PartialEq, Default)]
 pub struct Path {
     /// The node ID.
     pub id: NodeId,
     /// The span.
     pub span: Span,
+    /// Optional leading expression.
+    pub leading_expr: Option<Box<Expr>>,
```

Comment:
This may be a simplification that can be made: parsing paths whose leading term is obviously an expression and not just an ident into a separate AST node. My thought was that it should be a new AST node kind and not the Field kind so that we can preserve the fact that these are different syntaxes.

However, not all field accessor paths can be parsed like that. Some field accessors (and probably the most common kind) will just be a series of dot-separated idents, which to the parser looks identical to a namespace path, so the most common case will still be parsing field accessors into paths, just without the added complexity of having to deal with a leading expr node.

Reply:
I've made this change: parse the `leading_expr` paths into AST Field nodes instead and drop the `leading_expr` field from paths. I think that worked well to simplify a complicating factor to a lot of the logic.